projects/RabbitMQ.Client/client/impl/AsyncConsumerWorkService.cs

```diff
@@ -72,9 +99,22 @@ async Task Loop()
                         // Swallowing the task cancellation in case we are stopping work.
                     }
 
-                    while (_workQueue.TryDequeue(out Work work))
+                    if (_runInParallel)
+                    {
+                        while (_workBag.TryTake(out Work work))
+                        {
+                            _workTasks.Add(work.Execute(_model));
+                        }
+
+                        await Task.WhenAll(_workTasks).ConfigureAwait(false);
+                        _workTasks.Clear();
```

Comment:
I'm not entirely sure what this line does:
1) If there were no concurrent calls to `Enqueue`, the bag should be empty by now because we took everything;
2) If the `Enqueue` is called between the end of the while loop and this line, we can lose work items completely.